server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesEncoder.java

```diff
@@ -232,19 +273,27 @@ void decodeOrdinals(DataInput in, long[] out, int bitsPerOrd) throws IOException
         assert out.length == ES87TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE : out.length;
 
         long v1 = in.readVLong();
-        int header = (int) (v1 & 0b11L);
-        if (header == 0b00 || header == 0b10) {
-            // first bit is zero -> single run
+        if ((v1 & 0b1L) == 0) {
+            // first bit is 0 -> single run
             Arrays.fill(out, v1 >>> 1);
-        } else if (header == 0b01) {
+        } else if ((v1 & 0b11L) == 0b01) {
             // first two bits are 01 -> two runs
             v1 = v1 >>> 2;
             int runLen = in.readVInt();
             long v2 = v1 + in.readZLong();
             Arrays.fill(out, 0, runLen, v1);
             Arrays.fill(out, runLen, out.length, v2);
+        } else if ((v1 & 0b111L) == 0b111L) {
+            // first three bits are 111 -> cycle
+            int cycleLength = (int) v1 >>> 3;
+            for (int i = 0; i < cycleLength; i++) {
+                out[i] = in.readVLong();
+            }
+            for (int i = 0; i < out.length; i++) {
+                out[i] = out[i % cycleLength];
+            }
```

Comment:
Remainders can be a bit costly, do this instead? (It's possible that the compiler optimizes your loop, but the below loop doesn't rely on the compiler being too smart and isn't more complicated either?)

```java
for (int i = cycleLength; i < out.length; i++) {
    out[i] = out[i - cycleLength];
}
```

Alternatively, we could try to do more bulk copying by doing something like

```java
int length = cycleLength;
while (length < out.length) {
    int copyLength = Math.min(length, out.length - length);
    System.arraycopy(out, 0, out, length, copyLength);
    length += copyLength;
}
```